### PR TITLE
perf: optimize action configs processing

### PR DESCRIPTION
**Pull request by vvagaytsev**
**What this PR does / why we need it:**

* deduplicate paths when find minimal repo roots
* process action configs in batches

### CODE REVIEW

**Comment by vvagaytsev** [comment]
Closing in favor of #6535.

File: core/src/graph/actions.ts [hunk 1]
```diff
@@ -156,29 +169,34 @@ export const actionConfigsToGraph = profileAsync(async function actionConfigsToG
   const computedActionModes: { [key: string]: ComputedActionMode } = {}
 
   const preprocessActions = async (predicate: (config: ActionConfig) => boolean = () => true) => {
-    return await Promise.all(
-      Object.entries(configsByKey).map(async ([key, config]) => {
-        if (!predicate(config)) {
-          return
-        }
+    let batchNo = 1
```
Comment (stefreak) [inline core/src/graph/actions.ts:172]: minor nit: batchNo could be emitted by `sliceToBatches`

File: core/src/graph/actions.ts [hunk 2]
```diff
@@ -264,10 +282,14 @@ export const actionConfigsToGraph = profileAsync(async function actionConfigsToG
   log.debug(`Got ${preprocessedConfigs.length} action configs ${!!actionsFilter ? "with" : "without"} action filter`)
 
   // Optimize file scanning by avoiding unnecessarily broad scans when project is not in repo root.
-  const allPaths = preprocessedConfigs.map((c) => getSourcePath(c))
-  log.debug(`Finding minimal roots for ${allPaths.length} paths`)
+  const allPaths = new Set<string>()
```
Comment (stefreak) [inline core/src/graph/actions.ts:285]: Nice, good catch!

File: core/src/graph/actions.ts [hunk 3]
```diff
@@ -264,10 +282,14 @@ export const actionConfigsToGraph = profileAsync(async function actionConfigsToG
...
+  for (const preprocessedConfig of preprocessedConfigs) {
+    const sourcePath = getSourcePath(preprocessedConfig)
+    allPaths.add(sourcePath)
+  }
+  log.debug(`Finding minimal roots for ${allPaths.size} paths`)
   const minimalRoots = await garden.vcs.getMinimalRoots(log, allPaths)
-  log.debug(`Finding minimal roots for ${allPaths.length} paths`)
+  log.debug(`Found minimal roots for ${allPaths.size} paths`)
 
   // TODO: Maybe we could optimize resolving tree versions, avoid parallel scanning of the same directory etc.
```
Comment (stefreak) [inline core/src/graph/actions.ts:294]: does this PR maybe resolve this todo comment?

File: core/src/graph/actions.ts [hunk 4]
```diff
@@ -277,51 +299,57 @@ export const actionConfigsToGraph = profileAsync(async function actionConfigsToG
...
-          linkedSource,
-          remoteSourcePath,
-          supportedModes,
-          scanRoot: minimalRoots[getSourcePath(config)],
-        })
+  const actionConfigCount = Object.keys(preprocessResults).length
+  log.debug(`Processing ${actionConfigCount} action configs...`)
+  let batchNo = 1
+  for (const batch of sliceToBatches(preprocessResults, 100)) {
+    log.silly(`Processing actions batch #${batchNo} (${batch.length} items)`)
```
Comment (stefreak) [inline core/src/graph/actions.ts:306]: non-blocking: Might be nice to log total batches too here, if `sliceToBatches` emit that info

File: core/src/graph/solver.ts [hunk 5]
```diff
@@ -104,7 +104,7 @@ export class GraphSolver extends TypedEventEmitter<SolverEvents> {
 
     const log = this.log
     // TODO-0.13.1+: remove this lock and test with concurrent execution
-    return this.lock.acquire("solve", async () => {
+    // return this.lock.acquire("solve", async () => {
```
Comment (stefreak) [inline core/src/graph/solver.ts:107]: This change is not part of the PR description, was it meant to be made?

**Review by stefreak — DISMISSED** [review]
Great work overall! I added a few comments but if it was intentional to remove the solver lock here this looks great!

File: core/src/garden.ts [hunk 6]
```diff
@@ -872,15 +870,13 @@ export class Garden {
...
-            moduleConfig.plugin = provider.name
-            return this.addModuleConfig(moduleConfig)
-          })
-        )
-      )
+      for (const provider of providers) {
+        for (const moduleConfig of provider.moduleConfigs) {
+          // Make sure module and all nested entities are scoped to the plugin
+          moduleConfig.plugin = provider.name
+          this.addModuleConfig(moduleConfig)
```
Comment (stefreak) [inline core/src/garden.ts:877]: Great work! This is so much easier to read without the unnecessary async

File: core/src/graph/solver.ts (hunk 5 above)
Comment (vvagaytsev) [reply core/src/graph/solver.ts]: Good catch! That was done by accident, I have reverted that change.

**Review by stefreak — APPROVED** [review]

File: core/src/graph/actions.ts (hunk 4 above)
Comment (vvagaytsev) [reply core/src/graph/actions.ts:306]: Sure, will do 👍

File: core/src/graph/actions.ts (hunk 1 above)
Comment (vvagaytsev) [reply core/src/graph/actions.ts:172]: Yep, will do 👍

Comment (vvagaytsev) [reply core/src/graph/actions.ts:172]: Got some troubles with types, will do that in a separate PR later.

File: core/src/graph/actions.ts (hunk 3 above)
Comment (vvagaytsev) [reply core/src/graph/actions.ts:294]: I think, it does.

Comment (stefreak) [reply core/src/graph/actions.ts:294]: then todo should be removed

Comment (stefreak) [reply core/src/graph/actions.ts:294]:
```suggestion
```
